backend/handlers/math/outcomes/dbpm/marketshares_test.go

```diff
@@ -1,551 +1,719 @@
 	testcases := []struct {
-		Name          string
-		F_YES         float64
-		F_NO          float64
-		ExpectedF_YES float64
-		ExpectedF_NO  float64
+		Name           string
+		Bets           []models.Bet
+		ScaledPayouts  []int64
+		ExpectedExcess int64
 	}{
 		{
-			Name:          "PreventSimultaneousSharesHeld",
-			F_YES:         5.000000000000001, // Actual output from function
-			F_NO:          5.714285714285713, // Actual output from function
-			ExpectedF_YES: 5.000000,
-			ExpectedF_NO:  5.714286,
+			Name:           "InitialMarketState",
+			Bets:           []models.Bet{},
+			ScaledPayouts:  []int64{},
+			ExpectedExcess: 0, // No bets, no excess
 		},
 		{
-			Name:          "InfinityAvoidance",
-			F_YES:         0,
-			F_NO:          2,
-			ExpectedF_YES: 0,
-			ExpectedF_NO:  2,
+			Name: "FirstBetNoDirection",
+			Bets: []models.Bet{
+				generateBet(20, "NO", "one", 1, 0),
+			},
+			ScaledPayouts:  []int64{0},
+			ExpectedExcess: -20,
+		},
+		{
+			Name: "SecondBetYesDirection",
+			Bets: []models.Bet{
+				generateBet(20, "NO", "one", 1, 0),
+				generateBet(10, "YES", "two", 1, time.Minute),
+			},
+			ScaledPayouts:  []int64{19, 0}, // Scaled payouts < market volume
+			ExpectedExcess: -11,            // marketVolume = 30, scaledPayouts = 19
+		},
+		{
+			Name: "ThirdBetYesDirection",
+			Bets: []models.Bet{
+				generateBet(20, "NO", "one", 1, 0),
+				generateBet(10, "YES", "two", 1, time.Minute),
+				generateBet(10, "YES", "three", 1, 2*time.Minute),
+			},
+			ScaledPayouts:  []int64{20, 20, 0},
+			ExpectedExcess: 0, // marketVolume = 40, scaledPayouts = 40
+		},
+		{
+			Name: "FourthBetNegativeNoDirection",
+			Bets: []models.Bet{
+				generateBet(20, "NO", "one", 1, 0),
+				generateBet(10, "YES", "two", 1, time.Minute),
+				generateBet(10, "YES", "three", 1, 2*time.Minute),
+				generateBet(-10, "NO", "one", 1, 3*time.Minute),
+			},
+			ScaledPayouts:  []int64{11, 13, 6, 0},
+			ExpectedExcess: 0, // marketVolume = 30, scaledPayouts = 30
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.Name, func(t *testing.T) {
+			actualExcess := calculateExcess(tc.Bets, tc.ScaledPayouts)
+			if actualExcess != tc.ExpectedExcess {
+				t.Errorf(
+					"Test %s failed: expected excess %d, got %d",
+					tc.Name, tc.ExpectedExcess, actualExcess,
+				)
+			}
+		})
+	}
+}
+
+// theoretically this test case should never occur.
+// that being said, we're testing deducting from newest to oldest
+func TestAdjustForPositiveExcess(t *testing.T) {
+	testcases := []struct {
+		Name           string
+		ScaledPayouts  []int64
+		Excess         int64
+		ExpectedResult []int64
+	}{
+		{
+			Name:           "NoExcess",
+			ScaledPayouts:  []int64{10, 20, 30},
+			Excess:         0,
+			ExpectedResult: []int64{10, 20, 30}, // No adjustment needed
+		},
+		{
+			Name:           "SmallExcess",
+			ScaledPayouts:  []int64{10, 20, 30},
+			Excess:         2,
+			ExpectedResult: []int64{10, 19, 29},
 		},
```

Comment:
Ok, added a test for 1000, now it's taking forever to run the tests, so I'm going to have to go back on what I just said in a recent comment about not being able to figure out the excess allocation logic because evidently even just dealing with 1000 is untenable.